### PR TITLE
[core] set relation registry in schema upon finalization

### PR DESCRIPTION
**Pull request by GustavoCaso**
@solnic @flash-gordon 

I created this PR so we can discuss this https://github.com/rom-rb/rom-sql/issues/247. The finalize process still quite complex and I'm sure I'm missing something.

For what it looks like at this point when this message `configuration.relations.registry.created` is trigger the `RealtionRegistry` has been created. 

So I thought might be a simple solution to set the relations for each schema at this point.

### CODE REVIEW

File: core/lib/rom/schema.rb [hunk 1]
```diff
@@ -42,6 +42,7 @@ class Schema
       registry.each do |_, relation|
         unless relation.schema.frozen?
           relation.schema.finalize_associations!(relations: registry)
+          relation.schema.finalize_relations!(relations: registry)
```
Comment (solnic) [inline core/lib/rom/schema.rb]: Instead of doing this, it'd be better to [tweak this line](https://github.com/rom-rb/rom/blob/master/core/lib/rom/setup/finalize/finalize_relations.rb#L106) and do:

``` ruby
options = { __registry__: registry, mappers: mappers, schema: schema.with(relations: registry), **plugin_options }
```

File: core/spec/integration/setup_spec.rb [hunk 2]
```diff
@@ -30,6 +30,13 @@
       expect(tasks.users).to eql(users)
       expect(tasks.users.commands).to be_kind_of(ROM::CommandRegistry)
     end
+
+    it 'configures rom schema to store relations' do
+      users_schema = users_relation.schema
+      tasks_schema = tasks_relation.schema
+      expect(users_schema.relations).to be_kind_of(ROM::RelationRegistry)
+      expect(tasks_schema.relations).to be_kind_of(ROM::RelationRegistry)
```
Comment (solnic) [inline core/spec/integration/setup_spec.rb]: To make this spec more robust, it'd be good to do:

``` ruby
expect(users_schema.relations[:users]).to eql(users_relation)
expect(users_schema.relations[:tasks]).to eql(tasks_relation)
```

**Comment by GustavoCaso** [comment]
@flash-gordon @solnic I updated the code so we can merge it and work on the rom-sql with support for accessing relations in projection/restriction blocks 😄

File: core/lib/rom/schema.rb [hunk 3]
```diff
@@ -131,6 +131,8 @@ def self.attributes(attributes, attr_class)
     def initialize(*)
       super
 
+      initialize_primary_key_names
```
Comment (solnic) [inline core/lib/rom/schema.rb]: why was this added?

Comment (GustavoCaso) [reply core/lib/rom/schema.rb]: @solnic looks like after `schema.with(relations: registry)` the schema it returns have not initialized `primary_keys_names` and we were getting some errors in different parts of the specs, maybe the problem does not lie in the schema initialize method but in another part of the code.

```ruby
ROM::SQL::MissingPrimaryKeyError:
        Missing primary key for :users
```
Once I added that line the build looks like its working, but maybe there is something else that I'm missing. I happy will keep looking for other solution.

**Comment by flash-gordon** [comment]
@solnic @GustavoCaso I force-pushed the branch, I made it take PK names from canonical schema, I think this should be fine for now

File: core/lib/rom/schema.rb [hunk 4]
```diff
@@ -416,6 +416,22 @@ def to_ast
...
+      end
+    end
+
+    def primary_key_names
+      if canonical?
+        @primary_key_names if defined? @primary_key_names
+      else
+        canonical.primary_key_name
+      end
+    end
```
Comment (solnic) [inline core/lib/rom/schema.rb]: we shouldn't have this type of logic at run-time :/ How about we make these part of options instead? This way they can be maintained across instances.

File: core/spec/integration/setup_spec.rb [hunk 5]
```diff
@@ -30,6 +30,13 @@
       expect(tasks.users).to eql(users)
       expect(tasks.users.commands).to be_kind_of(ROM::CommandRegistry)
     end
+
+    it 'configures rom schema to store relations' do
+      users_schema = users_relation.schema
+      tasks_schema = tasks_relation.schema
+      expect(users_schema.relations[:users]).to eql(users_relation)
+      expect(tasks_schema.relations[:tasks]).to eql(tasks_relation)
+    end
```
Comment (solnic) [inline core/spec/integration/setup_spec.rb]: So, I understand that this makes sure we have finalized registry accessible via schemas, but there's also this change related to primary keys, and I'm not sure how that's related. Can we / should we have a separate spec example that checks primary key methods? ie canonical vs projected?

File: core/lib/rom/schema.rb (hunk 4 above)
Comment (flash-gordon) [reply core/lib/rom/schema.rb]: couldn't we just memoize them instead? Passing them as options is also not required, strictly speaking, because they are not always needed. And yes, either memoization or passing them with options is better than using ivars
